File: src/algorithms/python/Best_Time_to_Buy_and_Sell_Stock_with_Cooldown.py
# -*- coding: utf-8 -*-


class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        if len(prices) <= 1:
            return 0
        if len(prices) == 2:
            return max(prices[1] - prices[0], 0)
        dp = [[0 for x in range(2)] for y in prices]
        dp[0][0], dp[0][1] = -prices[0], 0
        dp[1][0], dp[1][1] = (max(-prices[0], -prices[1]),
                              max(prices[1] - prices[0], 0))
        for i in range(2, len(prices)):
            dp[i][0] = max(dp[i - 1][0], dp[i - 2][1] - prices[i])
            dp[i][1] = max(dp[i - 1][1], dp[i - 1][0] + prices[i])
        return dp[len(prices) - 1][1]


if __name__ == "__main__":
    solution = Solution()
    print solution.maxProfit([1, 2, 3, 0, 2])

File: src/algorithms/cpp/Best_Time_to_Buy_and_Sell_Stock_with_Cooldown.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if (SZ(prices) <= 1)    return 0;
        vector <pair<int, int>> dp(SZ(prices), make_pair(0, 0));
        dp[0] = make_pair(- prices[0], 0);
        dp[1] = make_pair(max(- prices[0], - prices[1]), max(0, prices[1] - prices[0]));
        for (int i = 2; i < SZ(prices); ++ i) {
            dp[i].first = max(dp[i - 1].first, dp[i - 2].second - prices[i]);
            dp[i].second = max(dp[i - 1].second, dp[i - 1].first + prices[i]);
        }
        return dp[SZ(prices) - 1].second;
    }
};

int main() {
    return 0;
}
